pkg/pgmodel/migrate.go

```diff
@@ -58,101 +57,159 @@ func (t *mySrc) replaceSchemaNames(r io.ReadCloser) (io.ReadCloser, error) {
+	} else {
+		_, err = db.Exec(metadataUpdateNoExtension, key, value, true)
 	}
-	r, err = t.replaceSchemaNames(r)
 	return
 }
 
-func (t *mySrc) ReadDown(version uint) (r io.ReadCloser, identifier string, err error) {
-	r, identifier, err = t.Driver.ReadDown(version)
+func versionGet(db *sql.DB) (*semver.Version, error) {
+	res, err := db.Query(metadataGet, "version")
+
 	if err != nil {
-		return
+		return nil, err
 	}
-	r, err = t.replaceSchemaNames(r)
-	return
-}
 
-func metadataUpdate(db *sql.DB, withExtension bool, key string, value string) {
-	/* Ignore error if it doesn't work */
-	if withExtension {
-		_, _ = db.Exec(metadataUpdateWithExtension, key, value, true)
-	} else {
-		_, _ = db.Exec(metadataUpdateNoExtension, key, value, true)
+	if !res.Next() {
+		// No version data is a fresh install.
+		return nil, nil
+	}
+
+	var version semver.Version
+
+	err = res.Scan(&version)
+
+	if err != nil {
+		return nil, err
 	}
+
+	return &version, nil
 }
 
 var migrateMutex = &sync.Mutex{}
 
-// Migrate performs a database migration to the latest version
-func Migrate(db *sql.DB, versionInfo VersionInfo) (err error) {
-	migrateMutex.Lock()
-	defer migrateMutex.Unlock()
-	// The migration table will be put in the public schema not in any of our schema because we never want to drop it and
-	// our scripts and our last down script drops our shemas
-	driver, err := postgres.WithInstance(db, &postgres.Config{MigrationsTable: "prom_schema_migrations"})
+func execMigrationFiles(tx *sql.Tx, name string) error {
+	f, err := migrations.MigrationFiles.Open(name)
 	if err != nil {
-		return fmt.Errorf("cannot create driver due to %w", err)
+		return fmt.Errorf("unable to get migration scripts: name %s, err %w", name, err)
 	}
 
-	_, err = db.Exec(timescaleInstall)
+	contents, err := replaceSchemaNames(f)
 	if err != nil {
-		return fmt.Errorf("timescaledb failed to install due to %w", err)
+		return fmt.Errorf("unable to read preinstall scripts: name %s, err %w", name, err)
 	}
 
-	src, err := httpfs.New(migrations.SqlFiles, "/")
+	_, err = tx.Exec(string(contents))
 	if err != nil {
-		return err
+		return fmt.Errorf("error executing preinstall scripts: name %s, err %w", name, err)
 	}
-	src = &mySrc{src}
+	return nil
+}
 
-	m, err := migrate.NewWithInstance("SqlFiles", src, "Postgresql", driver)
+func installVersionScripts(tx *sql.Tx, dbVersion, appVersion *semver.Version) error {
+	f, err := migrations.MigrationFiles.Open(versionScripts)
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to open migration scripts: %w", err)
 	}
-	defer func() {
-		sourceErr, databaseErr := m.Close()
-		//don't override error if already set
+
+	entries, err := f.Readdir(-1)
+	if err != nil {
+		return fmt.Errorf("unable to get migration scripts: %w", err)
+	}
+
+	sort.Slice(entries, func(i, j int) bool {
+		first, err := semver.Make(entries[i].Name())
+		if err != nil {
+			return false
+		}
+		second, err := semver.Make(entries[i].Name())
```

Comment:
BUG: second also has index `i`. So both first and second will be the same thing.